compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1115,51 +1118,40 @@ pub fn typeid_for_instance<'tcx>(
         instance.args = strip_receiver_auto(tcx, instance.args)
     }
 
+    if let Some(impl_id) = tcx.impl_of_method(instance.def_id())
+        && let Some(trait_ref) = tcx.impl_trait_ref(impl_id)
+    {
+        // Trait methods will have a Self polymorphic parameter, where the concreteized
+        // implementatation will not. We need to walk back to the more general trait method
+        let trait_ref = tcx.instantiate_and_normalize_erasing_regions(
+            instance.args,
+            ty::ParamEnv::reveal_all(),
+            trait_ref,
+        );
+        let invoke_ty = trait_object_ty(tcx, ty::Binder::dummy(trait_ref));
+        let method_id = tcx
+            .associated_item(instance.def_id())
+            .trait_item_def_id
+            .expect("Part of a trait implementation, but not linked to the def_id?");
+        // At the call site, any call to this concrete function through a vtable will be
+        // `Virtual(method_id, idx)` with appropriate arguments for the method. Since we have the
+        // original method id, and we've recovered the trait arguments, we can make the callee
+        // instance we're computing the alias set for match the caller instance.
+        //
+        // Right now, our code ignores the vtable index everywhere, so we use 0 as a placeholder.
+        // If we ever *do* start encoding the vtable index, we will need to generate an alias set
+        // based on which vtables we are putting this method into, as there will be more than one
+        // index value when supertraits are involved.
+        instance.def = ty::InstanceDef::Virtual(method_id, 0);
```

Comment:
Also, this is stripping off generic arguments on the method itself. That seems problematic. This method is called for non-vtable methods, no?

I think you might want to be using rebase_args to move the args from, and instead of using InstanceDef::Virtual, just use InstanceDef::Item with the new (trait) method def id you acquired above.

Reply:
Ah yeah, this would get called on non-vtable methods since I'm not using `CfiShim` anymore (which was only getting put in where the adjustment was needed).

Do you think that I should be using `rebase_args` and doing the transformation unctionditionally, or should I be trying to check if the method is vtable-safe before transforming it, to keep non-vtable methods from getting a `dyn` signature?

Reply:
Could you do both? 😂

Object types only have ReErased early-bound lifetimes, but making CFI do too much work is better than not doing enough

And also only checking for object safe methods is a good way to limit the fallout of that shortcoming you elaborated above.

Ping me on zulip if you need help w the arg rebasing, it can be kind of annoying